### DISCUSSION TITLE
Tired : A new framework : Unified UI (Uniui)

### DISCUSSION THREAD

**Discussion by Xyncgas** [dotnet/maui#17015] [General]
https://github.com/dotnet/maui/discussions/15203#discussioncomment-6827156

Today I was browsing the discussion, happened to written the idea down which is basically :

Let's retire MAUI, it has failed, no one is using it not even microsoft themselves, let's do something vastly different, by building our own browser instead of using XAML like legacy technologies, we can design the browser the way I described in a comment I linked above, let's call this new framework Uniui

**Comment by ivanjx** [comment]
maui has not succeeded yet because the usage is very small and not a lot of open source components for it yet unlike the web frameworks. i would like microsoft to SERIOUSLY investing into BLAZOR HYBRID. that would be enough for 99.99% of our use cases. creating something new from scratch again wont make the situation better.

https://xkcd.com/927/

**Reply by Xyncgas** [reply]
But in my opinion Blazor and MAUI didn't get it right

Because I am using a browser when I am using blazor, while MAUI/Xamarin frameworks are letting people write cross platform UI in a stupid XMAL language and has a lot of bugs and different to the way people write UI on browser with html and CSS

So I say, making a browser ourselves to reinvent the wheel which is necessary because we are going to make it a little different.

You're going to write UI using css-styling and using Javascript / C# / python / whatever language you choose, and how is that possible because we are going to be calling some binary interface from libraries in the developer environments you use, internally the binary interface is going to be efficient at least better than the current way for .NET telling the webview process what to do

So there is the performance benefit of writing UI in whatever language you'd like, and the benefit for building-styling UI the way it was on the web, simply because we are making our own browser, we can embed the browser anywhere including in the actual browser we are using through javascript or standardized to be put directly in every browser like webassembly is today, we can embed the browser on desktop, we can embed the browser on the server and ask the server to emit rpc to render UI on remote devices that's running the browser we can ask the server to emit static html and cache it on CDN and that would be generated say by the UI code you've written in python or c++. Now it is rather like a framework called Uniui than a browser called Uniui and that's how things should be

What'd you say

**Reply by ivanjx** [reply]
creating a whole browser-like platform is unfortunately isnt what microsoft is good at (IE, Legacy Edge). let alone adding features like that.

**Reply by Xyncgas** [reply]
Just require talented people to do it

**Comment by true-perfect-code** [comment]
I agree that MAUI Blazor is the future.

XAML is another Microsoft story where they are trying to reinvent the wheel. I am therefore sure that it would look much better for Microsoft if they were to break away from XAML. However, this is difficult, but there is nothing to stop Microsoft from pushing Blazor, which (at least it seems to me) is the case.

@Xyncgas 
At the moment, the most important thing is that we NET developers don't lose faith and stick with it, otherwise Microsoft's efforts will be for nothing!

**Reply by Xyncgas** [reply]
Blazor has to compile a dom on every render and update the dom on the browser, wouldn't it be nicer to directly tell the browser which element needs to do what while having the structure and safety you get via a maintained state as the virtual dom

We can do the optimization by removing the boundary between webview and our .NET environment, by directly running .NET inside this browser we built not through webassembly but directly and it can manipulate the UI not through javascript / html (even webassembly has to manipulate ui this way) but directly move the pixels, now we can build UI that runs on this browser that .NET is directly running inside, and the UI built this way would have big performance and big capability, we can draw and we can still bring React like framework back and we can animate 120 fps / sec and ray traced our buttons

**Reply by ivanjx** [reply]
the performance of blazor is severely limited because it has to pass through the js interop to update the dom. there is a proposal on the web standard to enable this feature so that wasm can access the dom tree directly but i dont think it will come any time soon. we will just have to wait.

**Reply by Xyncgas** [reply]
would be better for browsers to support both the DOM and Canvas like UI together to render a website, because a we are presenting not documents but UI nowadays, so would be straight forward to start using pixels to paint app's UI and we can make them compatible so UI designed this way can still be translated back to html + css + javascript, but we get the most performance running in this binary mode because we don't have to maintain and update a DOM but there is always going to be some kinds of model in the computer's memory for the browser to put the pixels on screen it's just now the model's more efficient than the Document Object Model existing on the major browsers today

The opinions I'm having changes if blazor start supporting highly-efficient way of 3D rendering, manipulating pixels in dotnet directly and having easy access to the capability of different graphics cards

**Comment by MhAllan** [comment]
@true-perfect-code  how is MAUI Blazor is different from Website Blazor? just a "saved blazor"? and we deal with css and the heck of frontend :D

**Reply by ivanjx** [reply]
maui blazor c# will run natively while blazor wasm c# will run on top of wasm. features like multithreading wont work on wasm (yet).

**Reply by Xyncgas** [reply]
At least you can implement UI easier in browser because low level features are abstracted properly, whereas in Xamarin related technology there is a gap between scripting in XAML to actually getting things done like implementing a right clicked context menu then showing another important element at the center of the screen, all I am suggesting is making the browser part of the framework we use so we can use it in different programming languages and the way we are manipulating UI doesn't have to be plain-text html / css and eventually we can trans-compile it to html + css + js during our compilation or later when the software is running

**Reply by Xyncgas** [reply]
Or maybe we let AI take our job, writing our software within unity3D

**Reply by MhAllan** [reply]
@Xyncgas  You implement a UI that doesn't have anything native, the keyboard will chock until it opens, the camera, ok try to give it permissions, the listing, virtualize yourself!  .. it will be a web page app, which is terrible. 

@ivanjx, do you think all these websites that are slow and glitchy on you phone is because JS? the problem is not JS. it is that native controls have UI acceleration, a list of images on a web page will never be like it on a native listview.

**Reply by Xyncgas** [reply]
Browser can be GPU accelerated
meanwhile I discovered the UNO platform is kinda a browser

**Reply by Stedy59** [reply]
@Xyncgas Are you using Maui? Or something else...
